doc/tutorial/tutorial.md

```diff
@@ -1,40 +1,650 @@
+```
+
+You should obtain the following output:
+
+```
+NetHostString{1,"Host: IP=170.187.204.221, MAC=11:22:33:44:55:66"}
+NetHostString{2,"Host: IP=160.176.192.208, MAC=10:20:30:40:50:60"}
+```
+
+### Bit vectors
+
+The design of DDlog was strongly influenced by applications from the domain of networking, where
+programmers frequently deal with bit vectors of non-standard sizes.  This motivated DDlog's
+bit-vector type.  We have encountered several instances of this type, e.g., `typedef UUID=bit<128>`.
+In addition to standard arithmetic and bitwise operators over bit vectors (see [language
+reference](../language_reference/language_reference.md#expressions)), DDlog supports bit vector
+slicing and concatenation.  Slicing is expressed using the `x[N:M]` syntax examplified in the
+previous section.  Bit vector concatenation is implemented by the `++` operator (which is also
+overloaded for strings).  Concatenation can be applied to bit vectors of arbitrary width, producing
+a bit vector whose width is the sum of the widths of its arguments.  The following example
+illustrates both operators:
+
+```
+// Form IP address from bytes using bit vector concatenation
+function ip_from_bytes(b3: bit<8>, b2: bit<8>, b1: bit<8>, b0: bit<8>)
+    : ip_addr_t =
+{
+    IPAddr{.addr = b3 ++ b2 ++ b1 ++ b0}
+}
+
+// Check for multicast IP address using bit slicing
+function is_multicast_addr(ip: ip_addr_t): bool = ip.addr[31:28] == 14
+
+input relation Bytes(b3: bit<8>, b2: bit<8>, b1: bit<8>, b0: bit<8>)
+
+// convert bytes to IP addresses
+relation Address(addr: ip_addr_t)
+Address(ip_from_bytes(b3,b2,b1,b0)) :- Bytes(b3,b2,b1,b0).
+
+// filter multicast IP addresses
+relation MCastAddress(addr: ip_addr_t)
+MCastAddress(a) :- Address(a), is_multicast_addr(a).
+```
+
+DDlog offers two ways to write bit vector literals. First, they can be written as decimal numbers.
+DDlog performs *type inference* to determine whether the number should be interpreted as an unbounded
+mathematical integer or a bit vector.  In the latter case it also infers its bit width.  For
+example, the literal `125` is interpreted as a mathematical integer in `var x: int = 125` and as an
+8-bit bit-vector in `var x: bit<8> = 125`.
+
+Second, bit vector literals can be written with explicit width and radix specifiers using the
+`[<width>]<radix_specifier><value>` syntax, where `<radix>` is one of `'d`, `'h`, `'o`, `'b` for
+decimal, hexadecimal, octal and binary numbers respectively:
+
+```
+32'd15              // 32-bit decimal
+48'haabbccddeeff    // 48-bit hexadecimal
+3'b101              // 3-bit binary
+'d15                // bit vector whose value is 15 and whose width is determined from the context
+```
 
 ### Control flow and variables
 
-*if-then-else, sequencing, assignments, var*
-
-## Procedures
-
-### Extern procedures
+Complex computations can be expressed in DDlog using control-flow constructs:
+
+1. *Sequencing* evaluates semicolumn-separated expressions in order.
```

Comment:
A series of expressions separated by semicolons are evaluated in the order they appear in the program.
semicolons, not semicolumns.